File: src/sdp/validators/ws_completion/verifier.py
"""Workstream completion verifier - validates WS is actually complete with evidence."""

from pathlib import Path
from typing import Any

from sdp.validators.ws_completion.checkers import (
    extract_coverage_value,
    verify_commands as check_commands,
    verify_coverage as check_coverage,
    verify_output_files as check_output_files,
)
from sdp.validators.ws_completion.models import CheckResult, VerificationResult
from sdp.validators.ws_completion.parser import find_ws_file, parse_ws_file


class WSCompletionVerifier:
    """Verify workstream completion with evidence."""

    def __init__(self, ws_dir: Path = Path("docs/workstreams")):
        """Initialize verifier.

        Args:
            ws_dir: Base directory for workstream files
        """
        self.ws_dir = ws_dir

    def verify_output_files(self, ws_data: dict[str, Any]) -> list[CheckResult]:
        """Check all output files in scope exist.

        Args:
            ws_data: Parsed workstream data

        Returns:
            List of check results for each file
        """
        return check_output_files(ws_data)

    def verify_commands(self, ws_data: dict[str, Any]) -> list[CheckResult]:
        """Run verification commands and check exit codes.

        Args:
            ws_data: Parsed workstream data

        Returns:
            List of check results for each command
        """
        return check_commands(ws_data)

    def verify_coverage(self, ws_data: dict[str, Any]) -> CheckResult | None:
        """Check test coverage meets threshold.

        Args:
            ws_data: Parsed workstream data

        Returns:
            Check result or None if no coverage requirement
        """
        return check_coverage(ws_data)

    def _find_ws_file(self, ws_id: str) -> Path | None:
        """Find WS file by ID (backward compatibility).

        Args:
            ws_id: Workstream ID

        Returns:
            Path to WS file or None
        """
        return find_ws_file(ws_id, self.ws_dir)

    def _parse_coverage_from_output(self, output: str) -> float | None:
        """Parse coverage percentage from pytest output (backward compatibility).

        Args:
            output: pytest stdout

        Returns:
            Coverage percentage or None
        """
        from sdp.validators.ws_completion.checkers import _parse_coverage_from_output

        return _parse_coverage_from_output(output)

    def _extract_coverage(self, check: CheckResult | None) -> float | None:
        """Extract coverage value from check result (backward compatibility).

        Args:
            check: Coverage check result

        Returns:
            Coverage percentage or None
        """
        return extract_coverage_value(check)

    def _parse_ws_file(self, ws_path: Path) -> dict[str, Any]:
        """Parse WS file for verification data (backward compatibility).

        Args:
            ws_path: Path to WS file

        Returns:
            Dict with scope_files, verification_commands, etc.
        """
        return parse_ws_file(ws_path)


    def verify(self, ws_id: str) -> VerificationResult:
        """Run all verification checks.

        Checks:
        1. All scope_files output exist
        2. All Verification commands pass
        3. Coverage meets threshold
        4. AC checkboxes accurate

        Args:
            ws_id: Workstream ID (e.g., "00-032-26")

        Returns:
            VerificationResult with all check results
        """
        checks: list[CheckResult] = []
        missing_files: list[str] = []
        failed_commands: list[str] = []

        # Find WS file
        ws_path = find_ws_file(ws_id, self.ws_dir)
        if not ws_path:
            return VerificationResult(
                ws_id=ws_id,
                passed=False,
                checks=[
                    CheckResult(
                        name="Find WS",
                        passed=False,
                        message=f"Workstream file not found: {ws_id}",
                        evidence=None,
                    )
                ],
                coverage_actual=None,
                missing_files=[],
                failed_commands=[],
            )

        # Parse WS file
        ws_data = parse_ws_file(ws_path)

        # Check 1: Verify output files exist
        file_checks = self.verify_output_files(ws_data)
        checks.extend(file_checks)
        missing_files = [c.message for c in file_checks if not c.passed]

        # Check 2: Run verification commands
        cmd_checks = self.verify_commands(ws_data)
        checks.extend(cmd_checks)
        failed_commands = [c.name for c in cmd_checks if not c.passed]

        # Check 3: Verify coverage
        coverage_check = self.verify_coverage(ws_data)
        if coverage_check:
            checks.append(coverage_check)

        # Determine overall pass/fail
        passed = all(c.passed for c in checks)

        return VerificationResult(
            ws_id=ws_id,
            passed=passed,
            checks=checks,
            coverage_actual=extract_coverage_value(coverage_check) if coverage_check else None,
            missing_files=missing_files,
            failed_commands=failed_commands,
        )

File: sdp-plugin/internal/verify/verifier.go
package verify

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/fall-out-bug/sdp/internal/config"
	"github.com/fall-out-bug/sdp/internal/quality"
	"github.com/fall-out-bug/sdp/internal/security"
)

// Verifier handles workstream completion verification
type Verifier struct {
	parser *Parser
}

// NewVerifier creates a new workstream verifier
func NewVerifier(wsDir string) *Verifier {
	return &Verifier{
		parser: NewParser(wsDir),
	}
}

// VerifyOutputFiles checks all scope_files exist and are within project root (path traversal safety).
func (v *Verifier) VerifyOutputFiles(wsData *WorkstreamData) []CheckResult {
	checks := make([]CheckResult, 0, len(wsData.ScopeFiles))
	projectRoot, rootErr := config.FindProjectRoot()
	if rootErr != nil {
		projectRoot = ""
	}

	for _, filePath := range wsData.ScopeFiles {
		check := CheckResult{
			Name: fmt.Sprintf("File: %s", filePath),
		}

		// Path traversal: ensure path is within project root
		if projectRoot != "" {
			if err := security.ValidatePathInDirectory(projectRoot, filePath); err != nil {
				check.Passed = false
				check.Message = fmt.Sprintf("Path outside project: %v", err)
				checks = append(checks, check)
				continue
			}
		}

		// Check if file exists
		if _, err := os.Stat(filePath); err == nil {
			// File exists
			check.Passed = true
			check.Message = filePath
			absPath, err := filepath.Abs(filePath)
			if err != nil {
				slog.Debug("filepath.Abs failed, using relative path", "path", filePath, "error", err)
				absPath = filePath // Fall back to original path
			}
			check.Evidence = absPath
		} else {
			// File doesn't exist
			check.Passed = false
			check.Message = fmt.Sprintf("Missing: %s", filePath)
		}

		checks = append(checks, check)
	}

	return checks
}

// VerifyCommands runs verification commands with security validation (sdp-5ho2).
// Derives per-command timeouts from the parent context; respects ctx cancellation for graceful shutdown.
func (v *Verifier) VerifyCommands(ctx context.Context, wsData *WorkstreamData) []CheckResult {
	checks := []CheckResult{}

	for _, cmd := range wsData.VerificationCommands {
		check := CheckResult{
			Name: fmt.Sprintf("Command: %s", truncate(cmd, 50)),
		}

		cmdParts := strings.Fields(cmd)
		if len(cmdParts) == 0 {
			check.Passed = false
			check.Message = "Empty command"
			checks = append(checks, check)
			continue
		}

		timeout := verificationTimeout()
		cmdCtx, cancel := context.WithTimeout(ctx, timeout)

		command, err := security.SafeCommand(cmdCtx, cmdParts[0], cmdParts[1:]...)
		if err != nil {
			cancel()
			check.Passed = false
			check.Message = fmt.Sprintf("Security validation: %v", err)
			checks = append(checks, check)
			continue
		}

		output, err := command.CombinedOutput()
		cancel()

		if err != nil {
			check.Passed = false
			check.Message = fmt.Sprintf("Exit code: %v", err)
			check.Evidence = truncate(string(output), 500)
		} else {
			check.Passed = true
			check.Message = "Exit code: 0"
			check.Evidence = truncate(string(output), 500)
		}

		checks = append(checks, check)
	}

	return checks
}

// VerifyCoverage runs actual coverage check via quality.Checker. ctx is used for cancellation.
func (v *Verifier) VerifyCoverage(ctx context.Context, wsData *WorkstreamData) *CheckResult {
	if wsData.CoverageThreshold == 0 {
		return nil
	}

	projectRoot, err := config.FindProjectRoot()
	if err != nil {
		return &CheckResult{
			Name:    "Coverage Check",
			Passed:  false,
			Message: fmt.Sprintf("project root: %v", err),
		}
	}

	checker, err := quality.NewChecker(projectRoot)
	if err != nil {
		return &CheckResult{
			Name:    "Coverage Check",
			Passed:  false,
			Message: fmt.Sprintf("checker init: %v", err),
		}
	}

	result, err := checker.CheckCoverage(ctx)
	if err != nil {
		return &CheckResult{
			Name:    "Coverage Check",
			Passed:  false,
			Message: fmt.Sprintf("coverage check: %v", err),
		}
	}

	threshold := result.Threshold
	if wsData.CoverageThreshold > 0 {
		threshold = wsData.CoverageThreshold
	}
	passed := result.Coverage >= threshold

	return &CheckResult{
		Name:     "Coverage Check",
		Passed:   passed,
		Message:  fmt.Sprintf("Coverage: %.1f%% (threshold: %.1f%%)", result.Coverage, threshold),
		Evidence: result.Report,
	}
}

// Verify runs all verification checks. ctx is used for command timeouts and cancellation.
func (v *Verifier) Verify(ctx context.Context, wsID string) *VerificationResult {
	start := time.Now()

	result := &VerificationResult{
		WSID:           wsID,
		Checks:         []CheckResult{},
		MissingFiles:   []string{},
		FailedCommands: []string{},
	}

	if ctx == nil {
		ctx = context.Background()
	}

	// Find WS file
	wsPath, err := v.parser.FindWSFile(wsID)
	if err != nil {
		result.Passed = false
		result.Checks = append(result.Checks, CheckResult{
			Name:    "Find WS",
			Passed:  false,
			Message: err.Error(),
		})
		result.Duration = time.Since(start)
		return result
	}

	// Parse WS file
	wsData, parseErr := v.parser.ParseWSFile(wsPath)
	if parseErr != nil {
		result.Passed = false
		result.Checks = append(result.Checks, CheckResult{
			Name:    "Parse WS",
			Passed:  false,
			Message: parseErr.Error(),
		})
		result.Duration = time.Since(start)
		return result
	}

	// Check 1: Verify output files
	fileChecks := v.VerifyOutputFiles(wsData)
	result.Checks = append(result.Checks, fileChecks...)
	for _, check := range fileChecks {
		if !check.Passed {
			result.MissingFiles = append(result.MissingFiles, check.Message)
		}
	}

	// Check 2: Run verification commands
	cmdChecks := v.VerifyCommands(ctx, wsData)
	result.Checks = append(result.Checks, cmdChecks...)
	for _, check := range cmdChecks {
		if !check.Passed {
			result.FailedCommands = append(result.FailedCommands, check.Name)
		}
	}

	// Check 3: Verify coverage
	coverageCheck := v.VerifyCoverage(ctx, wsData)
	if coverageCheck != nil {
		result.Checks = append(result.Checks, *coverageCheck)
	}

	// Determine overall pass/fail
	result.Passed = true
	for _, check := range result.Checks {
		if !check.Passed {
			result.Passed = false
			break
		}
	}

	result.Duration = time.Since(start)
	return result
}

// verificationTimeout returns verification command timeout from config (or env, or default).
func verificationTimeout() time.Duration {
	root, err := config.FindProjectRoot()
	if err != nil {
		return config.TimeoutFromEnv("SDP_TIMEOUT_VERIFICATION", 60*time.Second)
	}
	cfg, err := config.Load(root)
	if err != nil || cfg == nil {
		return config.TimeoutFromEnv("SDP_TIMEOUT_VERIFICATION", 60*time.Second)
	}
	return config.TimeoutFromConfigOrEnv(cfg.Timeouts.VerificationCommand, "SDP_TIMEOUT_VERIFICATION", 60*time.Second)
}

// truncate truncates a string to max length
func truncate(s string, maxLen int) string {
	if len(s) <= maxLen {
		return s
	}
	return s[:maxLen] + "..."
}
